### ISSUE TITLE
kvs: add API for RFC 18 KVS event logs

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#1654]
A standardized format for timestamped KVS event logs was proposed in [RFC 18](https://github.com/flux-framework/rfc/blob/master/spec_18.adoc).  We should consider implementing an API for this.

Straw man:
```c
// Append an event to log specified by key using current wallclock as timestamp
flux_future_t *flux_kvs_evlog_append (flux_t *h, const char *key,
                                      const char *name, const char *context);

// Retrieve current event log
// If flags & FLUX_KVS_WATCH, future is fulfilled with a complete copy of the log each time it is updated
flux_future_t *flux_kvs_evlog_lookup (flux_t *h, const char *key, int flags);

// Get first/next event in log and optionally timestamp and context
// Cursor remains valid across multiple responses on the same future
const char *flux_kvs_evlog_lookup_get_first (flux_future_t *f,
                                             time_t *timestamp, const char **context);
const char *flux_kvs_evlog_lookup_get_next (flux_future_t *f,
                                            time_t *timestamp, const char **context);
```

**Comment by SteVwonder** [comment]
The timing on this is impeccable.  @dongahn and I were just talking about this.

@dongahn: I confirmed with @garlick that the scheduler (and other modules) can append to their own individual event log by changing the `key` passed to `flux_kvs_evlog_append`.

**Comment by grondo** [comment]
I would argue for keeping the `eventlog` name in the API, or renaming event logs to "evlogs". `evlog` only saves 3 characters...

The API might be more flexible if a `flux_eventlog_t` handle was retrieved from the `flux_future_t` returned from `flux_kvs_evlog_lookup()`, then it could have iterators with a nicer interface and type checking. Also, for unit tests, a `flux_eventlog_t` could be created from a file and the API exercised that way. The future returned by lookup might also be used to "udpate" an eventlog_t.

Main consumer use-cases of the eventlog interface will be to get notified when an event occurs, or search for a given state and its timestamp, etc. It would be a nicer interface if these high level functions didn't take a future..

**Comment by dongahn** [comment]
> @dongahn: I confirmed with @garlick that the scheduler (and other modules) can append to their own individual event log by changing the key passed to flux_kvs_evlog_append.

It is not directly related to this, but as you and I chatted yesterday. For the main scheduler, I really do think that we need to hide the underlying mechanisms by which the states are being traced into a higher level API. 

The main scheduler should be able to make use of jobs' state changes without having to know the details of where/how the states are being tracked. It would be best to write the scheduler loop service as a finite state machine that reacts to each and all the state changes it requires without having to know whether it needs to use eventing and/or watch and/or global state log and/or local state log.

This way, the complexity of having to deal with the underlying mechanism will stay within the implementation of the API while the scheduler loop module will stay lean and mean.

**Comment by garlick** [comment]
I think that's a good suggestion.  I'll iterate on the API in the description of this issue until we like it.

**Comment by garlick** [comment]
> It is not directly related to this, but as you and I chatted yesterday. For the main scheduler, I really do think that we need to hide the underlying mechanisms by which the states are being traced into a higher level API.

Yeah, and actually I think we will have a separate API to the job module for querying job states and events.  This will be a lower level interface used by the job module.
